src/storage/v2/replication/replication_client.hpp

```diff
@@ -121,32 +147,33 @@ class ReplicationStorageClient {
    * @param storage pointer to the storage associated with the client
    * @param gk gatekeeper access that protects the database; std::any to have separation between dbms and storage
    */
-  void StartTransactionReplication(uint64_t current_wal_seq_num, Storage *storage, DatabaseAccessProtector db_acc);
+  auto StartTransactionReplication(uint64_t current_wal_seq_num, Storage *storage, DatabaseAccessProtector db_acc)
+      -> std::optional<ReplicaStream>;
 
   // Replication clients can be removed at any point
   // so to avoid any complexity of checking if the client was removed whenever
   // we want to send part of transaction and to avoid adding some GC logic this
   // function will run a callback if, after previously calling
   // StartTransactionReplication, stream is created.
   template <InvocableWithStream F>
-  void IfStreamingTransaction(F &&callback) {
+  void IfStreamingTransaction(F &&callback, std::optional<ReplicaStream> &replica_stream) {
     // We can only check the state because it guarantees to be only
     // valid during a single transaction replication (if the assumption
     // that this and other transaction replication functions can only be
     // called from a one thread stands)
     if (State() != replication::ReplicaState::REPLICATING) {
       return;
     }
-    if (!replica_stream_ || replica_stream_->IsDefunct()) {
-      replica_state_.WithLock([this](auto &state) {
-        replica_stream_.reset();
+    if (!replica_stream || replica_stream->IsDefunct()) {
+      replica_state_.WithLock([&replica_stream](auto &state) {
+        replica_stream.reset();
         state = replication::ReplicaState::MAYBE_BEHIND;
       });
       LogRpcFailure();
       return;
     }
     try {
-      callback(*replica_stream_);  // failure state what if not streaming (std::nullopt)
+      callback(*replica_stream);  // failure state what if not streaming (std::nullopt)
```

Comment:
Does the catch need to reset?

Reply:
No need, we don't need to be so explicit with destruction anymore since it is not a private variable, will on RPC fail to change state to MAYBE_BEHIND, and destruction will happen in finalize transaction function, after the object gets out of scope